### PR TITLE
Extract comments to a separate table

### PR DESCRIPTION
**Pull request by licitdev**
## Scope

What's changed:

- A new `directus_comments` table for comments

## Potential Risks / Drawbacks

- Existing users reading into comments outside of the data studio will have to update their usage

## Review Notes / Questions

- Activity and revisions now exist for tracking
- Additional logic is added in the API to support legacy app queries to update comments
- A migration is required in a subsequent version to migrate legacy comments in `directus_activity` to `directus_comments`
- A legacy comment is only migrated into `directus_comments` when it is updated or deleted
- The legacy app usage can be tested by building the app from `main` and accessing it from the dev api server in this PR

---

Related to #17894

Closes SER-255

### CODE REVIEW

File: api/src/database/migrations/20240418A-separate-comments.ts
Comment (br41nslug) [inline api/src/database/migrations/20240418A-separate-comments.ts:1]: Do we have some form of timeout on migrations? if so then this loop may run into that for larger projects 🤔

File: api/src/services/graphql/index.ts
Comment (br41nslug) [inline api/src/services/graphql/index.ts:3188]: I think these should be updated to use the CommentsService rather than be deleted. We still want people to be able to create and update comments over graphql

File: sdk/src/rest/commands/delete/comments.ts
Comment (br41nslug) [inline sdk/src/rest/commands/delete/comments.ts:12]: Since the DELETE endpoint also supports updating by Query we should reflect that in 
```suggestion
	<Schema>(keysOrQuery: DirectusComment<Schema>['id'][] | Query<Schema, DirectusComment<Schema>>): RestCommand<void, Schema> =>
```

**Review by br41nslug — DISMISSED** [review]
This PR will require some updated to the API reference docs and a breaking change note to cover the updated API.

Not sure if this is an issue in my setup or this PR but i was getting some infinite getSchema loop errors while attempting to run bootstrap until i disabled CACHE_SCHEMA
Code_AQXxQUmlyd

File: api/src/database/migrations/20240418A-separate-comments.ts
Comment (licitdev) [reply api/src/database/migrations/20240418A-separate-comments.ts:1]: No, there isn't. CC @wrynegade

File: sdk/src/rest/commands/update/comments.ts
Comment (licitdev) [reply sdk/src/rest/commands/update/comments.ts]: Updated implementation via https://github.com/directus/directus/pull/22295/commits/33629c146738b2fdd766241dda1ea3ce65fd72e9

File: api/src/controllers/comments.ts
Comment (licitdev) [reply api/src/controllers/comments.ts]: No, it isn't. 👍 Removed in https://github.com/directus/directus/pull/22295/commits/e1053cb4a3e4628aca68d6b3d19f0809b358e941. Just as a note, it exists in many collections, which could be refactored in a separate PR.

image

File: api/src/services/graphql/index.ts
Comment (licitdev) [reply api/src/services/graphql/index.ts:3188]: The new `directus_comments` is usable via `/graphql/system` with a query like the following:

```gql
{
  comments (limit: 5) {
    id
    comment
    date_created
    user_created{
      first_name
    }
  }
}
```

File: sdk/src/rest/commands/delete/comments.ts
Comment (licitdev) [reply sdk/src/rest/commands/delete/comments.ts]: Updated via https://github.com/directus/directus/pull/22295/commits/40fe2953e86ecd9476805a79a6932a2fa47fc6b8

File: api/src/controllers/activity.ts
Comment (ComfortablyCoding) [inline api/src/controllers/activity.ts:69]: If we are leaving patch and delete for backwards compatible (per discussion with Ian) we should leave post as well. This would ensure all comment operation work while migration is taking place. We can remove these in a subsequent version once we are sure migrations are no longer a consideration

File: api/src/controllers/comments.ts
Comment (ComfortablyCoding) [inline api/src/controllers/comments.ts:76]: Following controller paradigms we should return null if no result
```suggestion
	res.locals['payload'] = { data: result || null, meta };
```

File: api/src/database/migrations/20240426A-separate-comments.ts
Comment (ComfortablyCoding) [inline api/src/database/migrations/20240426A-separate-comments.ts:69]: Should we be doing batch processing here as well same as we do for `up`?

File: api/src/services/comments.ts
Comment (ComfortablyCoding) [inline api/src/services/comments.ts:30]: Any reason why we do not user a schema here same as was done originally?

File: api/src/services/comments.ts
Comment (ComfortablyCoding) [inline api/src/services/comments.ts:36]: We should probably have this check for update as well, it would suffer from the same issues as create if left without it

File: api/src/services/graphql/index.ts
Comment (ComfortablyCoding) [inline api/src/services/graphql/index.ts:3197]: For the same reasoning as stated in https://github.com/directus/directus/pull/22295/files#r1677902904 we should leave these here for now

File: api/src/utils/get-service.ts
Comment (ComfortablyCoding) [inline api/src/utils/get-service.ts:35]: Is this an intentional change?

File: api/src/services/graphql/index.ts
Comment (ComfortablyCoding) [inline api/src/services/graphql/index.ts:2013]: Is this an intentional change?

File: app/src/views/private/components/comment-item-header.vue
Comment (ComfortablyCoding) [inline app/src/views/private/components/comment-item-header.vue:79]: This check seems redundant? The template would not render if `comment.user_created` does not exist. If this is the case we should be able to remove the v-else (private user) as well here as that should never occur either

**Comment by ComfortablyCoding** [comment]
> Should existing SDK & GQL remain unchanged until an upcoming version when the comment field is dropped from activity?

IMO SDK support can be dropped in this PR as we would start supporting the new API instantly. No reason it needs to be backwards compatible like the API is.

For GQL it should be inline with whatever we do for REST API

**Comment by ComfortablyCoding** [comment]
> Instead of removing the /activity/comment API, should an error message similar to the one in webhooks be added?

This should not be needed as in the initial release we are leaving in the logic for backwards compatibility reasons (which means we cannot error out) but for the most part documenting only the new API. We could do this in the subsequent release but I am in favour of just removing the endpoints altogether at that point.

**Comment by ComfortablyCoding** [comment]
We should be adding the documentation for the new comments endpoints and removing the ones for activity

**Comment by ComfortablyCoding** [comment]
Once this PR is merged we should update and merge https://github.com/directus/directus/pull/22962 as it is dependant/blocked by this one but should be done

File: api/src/utils/get-service.ts
Comment (ComfortablyCoding) [reply api/src/utils/get-service.ts:35]:
```suggestion
		 case 'directus_collections':
		 	return new CollectionsService(opts);
```

File: api/src/services/graphql/index.ts
Comment (ComfortablyCoding) [reply api/src/services/graphql/index.ts:2013]: This was unintentional per @licitdev , will add it back.

Comment (ComfortablyCoding) [reply api/src/services/graphql/index.ts:2097]: Upon checking this was an unused resolver, should be fine to leave it out

File: api/src/database/migrations/20240426A-separate-comments.ts
Comment (ComfortablyCoding) [reply api/src/database/migrations/20240426A-separate-comments.ts:61]: Resolved in 1f1c272e195cded9fe5ce0926781bcaf6e2d0777

File: api/src/utils/get-service.ts
Comment (ComfortablyCoding) [reply api/src/utils/get-service.ts]: Circling back here, it seems this was removed due to typing issues and not extending ItemsService. Seems we do not include any services that do not do so.

Removed again in 84311a846cd73321f9a3a695f3f12f49fd33f57c

File: app/src/views/private/components/comment-item-header.vue
Comment (ComfortablyCoding) [reply app/src/views/private/components/comment-item-header.vue]: Resolved in 7f1f2ea1d737046c480d563a93c97e885bc29518

File: api/src/services/comments.ts
Comment (ComfortablyCoding) [reply api/src/services/comments.ts:30]: Seems it does not play well with types, disregard

File: app/src/views/private/components/comments-sidebar-detail.vue
Comment (DanielBiegler) [inline app/src/views/private/components/comments-sidebar-detail.vue:122]: Passing in an array now changes the function signature, the type is now wrong for `loadUserPreviews` see line 246

File: api/src/controllers/activity.ts
Comment (DanielBiegler) [inline api/src/controllers/activity.ts:83]: Per the other comments, we seem to want backwards compatibility. Shouldnt we then not just use this existing activity controller to actually call the methods on the new `CommentsService`?

The migration moves everything to the comments table, why are we then still patching and deleting on the activity table instead of the comments table, the data is there? I'd say if someone then POSTs a new comment on `/activity/comment`, we ought to simply call `CommentsService.createOne` and thats it. The comments table should be the new source of truth right?

Then we also dont need two requests in `comments-sidebar-detail.vue`, we only need to fetch from the comments table. Pls correct me if I misunderstand something :pray:

Comment (licitdev) [reply api/src/controllers/activity.ts]: Great points! Yes to the above questions. The `CommentsService` is being refactored 👍

File: app/src/views/private/components/comments-sidebar-detail.vue
Comment (licitdev) [reply app/src/views/private/components/comments-sidebar-detail.vue]: Great eye, it looks like an error carried forward, where `Activity[]` was passed back then 👍 Updated in https://github.com/directus/directus/pull/22295/commits/02c1be310978c99cb5c056c74363c8474cf55563

File: api/src/controllers/activity.ts
Comment (ComfortablyCoding) [reply api/src/controllers/activity.ts]: I believe this originally done for the edge case where the migration was running and a comment was created.

Since the migration was refactored to only create the table this is most likely no longer a concern.

File: api/src/services/comments.ts
Comment (licitdev) [reply api/src/services/comments.ts]: I think the item access check should only be for create. For example, there could be a comments moderator who should be able to update or delete comments without needing to access the item.

Comment (licitdev) [reply api/src/services/comments.ts]: Updated the recommended permission to only allow for the updating of comments too. https://github.com/directus/directus/pull/22295/commits/0cd367d4f09df15c588b5f37fb578a23916847e0

File: api/src/services/comments.ts
Comment (ComfortablyCoding) [inline api/src/services/comments.ts:121]: This throws an `INTERNAL_SERVER_ERROR` response when the collection is non-existent. 

Server side response error is `Cannot find primary key for collection \"fake_collection\""` instead of the expected `You don't have permission to access collection \"article\" or it does not exist. Queried in root`

**Comment by ComfortablyCoding** [comment]
LGTM 🚀 

Some things to note:
1. For older comments the ID changes from a number (from revisions table) to a UUID (from comments table) on first update after API code changes. This can lead to confusion for subsequent updates if the user attempts to use the old (number) id a second time. Effect should be minimal as the app does account for this.
2. While https://github.com/directus/directus/pull/22295/commits/a12b49b54b0ded089bd66eaaa4ffb531eb6ab9d8 did fix the non-existent collection `Internal Server` error, this applies to non admin users only, for admin accounts this check is bypassed altogether and results in a `Foreign Key` error.

**Review by DanielBiegler — APPROVED** [review]
After the latest rounds of changes it worked for me, lets move on to the testing branch

**Pull request merged**
